test/freenet/client/filter/GIFFilterTest.java

```diff
@@ -0,0 +1,221 @@
+        "basn0g01.png",
+        // Truncated GIF file (share-the-safety-like.gif with removed terminator)
+        "share-the-safety-like.truncated.gif"
+    };
+
+    public void testKnownGood() {
+        for (String good : GOOD) {
+            assertEqualAfterFilter(good, good);
+        }
+    }
+
+    public void testFilterPairs() {
+        for (String[] pair : FILTER_PAIRS) {
+            assertEqualAfterFilter(pair[0], pair[1]);
+        }
+    }
+
+    public void testReject() {
+        for (String reject : REJECT) {
+            final InputStream inStream;
+            final NullOutputStream outStream;
+            try {
+                inStream = resourceToBucket(reject).getInputStream();
+                outStream = new NullOutputStream();
+            } catch (IOException e) {
+                throw new AssertionError(e);
+            }
+            ContentDataFilter filter = new GIFFilter();
+            try {
+                filter.readFilter(inStream, outStream, "", null, null);
+                fail("Filter did not fail on reject sample " + reject);
+            } catch (DataFilterException e) {
+                // Expected.
+            } catch (Exception e) {
+                throw new AssertionError("Unexpected exception in the content filter.", e);
+            }
+            try {
+                inStream.close();
+                outStream.close();
+            } catch (IOException e) {
+                throw new AssertionError(e);
+            }
+        }
+    }
+
+    /**
+     * Asserts that the test file in the first argument, after passing through the content filter,
+     * is equal to the reference file in the second argument.
+     *
+     * @param fileUnfiltered  the test file
+     * @param fileExpected    the reference file
+     */
+    private static void assertEqualAfterFilter(String fileUnfiltered, String fileExpected) {
+        Bucket input = resourceToBucket(fileUnfiltered);
+        Bucket expected = resourceToBucket(fileExpected);
+        Bucket filtered = filterGIF(input);
+        assertTrue("Filtered and expected output are not identical. " +
+            "Input = " + fileUnfiltered + ", expected = " + fileExpected,
+            equalBuckets(filtered, expected)
+        );
+    }
+
+    /**
+     * Checks for equality of Bucket contents.
+     */
+    private static boolean equalBuckets(Bucket a, Bucket b) {
+        try {
+            return Arrays.equals(BucketTools.toByteArray(a), BucketTools.toByteArray(b));
+        } catch (IOException e) {
+            throw new AssertionError(e);
+        }
+    }
+
+    /**
+     * Runs a Bucket through the content filter.
+     *
+     * @throws AssertionError on failure
+     */
+    private static Bucket filterGIF(Bucket input) {
+        ContentDataFilter filter = new GIFFilter();
+        Bucket output = new ArrayBucket();
+
+        InputStream inStream;
+        OutputStream outStream;
+        try {
```

Comment:
We’re on Java 7 now, we can use

    try (InputStream inStream = input.getInputStream();
        OutputStream outStream = output.getOutputStream()) {
        … do stuff …
    }

Also, in tests there is usually no need to catch exceptions, just let them bubble up so that the test method fails.